### PR TITLE
Show the lock screen placeholder on willResignActive.

### LINKED ISSUE #2026: When PIN enabled, backgrounding the app should immediately trigger the lockscreen.
### Steps to reproduce

1. Enable PIN
2. Start to background the app (e.g. by dragging up from the bottom of the screen)
3. Note that the app doesn't switch to lock screen.

#### What did you expect?

Expect the app screen to immediately switch to the lock screen as soon as the swipe begins, to stop folks from shouldersurfing contents in the task switcher, or screenshots of the app with sensitive content getting cached by the task switcher (and being used in the animation when the app is foregrounded, leaking content). The instant-hiding also makes the app feel more trustworthy with privacy.

If the app is fully backgrounded then when the app foregrounds again you'd expect to have to enter the PIN or FaceID.  But if the backgrounding is cancelled (i.e. the user stops dragging upwards) there's no need to re-auth the user.

#### What happened instead?

People shouldersurfing when i task switch in my app (or when i foreground the app, prior to unlocking it) will see leaked sensitive content.

### Your phone model

_No response_

### Operating system version

_No response_

### Application version

428

### Homeserver

_No response_

### Will you send logs?

No

### PR DESCRIPTION
**Pull request by pixlwave**
This PR fixes #2026. Also adds a couple of frames delay when window switching to avoid an issue where the transition would sometimes flicker through black.

First recording is background/foregrounding, the rest are various instances of system UI being shown that also triggers the placeholder as discussed in the issue.

https://github.com/vector-im/element-x-ios/assets/6060466/c07ae87a-c02c-45c6-b06a-58a9826ab947

**System UI**

https://github.com/vector-im/element-x-ios/assets/6060466/acd90a86-bd3c-49ae-a73c-dafb6c61b6fc

https://github.com/vector-im/element-x-ios/assets/6060466/6f658ae8-23f5-45b7-848a-8cb0d99e8f87

https://github.com/vector-im/element-x-ios/assets/6060466/793860ec-f815-4fba-b352-2bcba2bd6275

https://github.com/vector-im/element-x-ios/assets/6060466/892d1993-3958-4ea4-989b-600ac6d35c15

https://github.com/vector-im/element-x-ios/assets/6060466/5123a4bb-62f8-4752-b724-8eb96ccb6acb

### CODE REVIEW

**Review by alfogrillo — APPROVED** [review]
LGTM
(asked just one clarification question)

File: ElementX/Sources/Application/Windowing/WindowManager.swift [hunk 1]
```diff
@@ -58,21 +63,33 @@ class WindowManager {
...
-        overlayWindow.isHidden = false
-        alternateWindow.isHidden = true
+        switchTask = Task {
+            mainWindow.isHidden = false
+            overlayWindow.isHidden = false
+            
+            // Delay hiding to make sure the main windows are visible.
+            try await Task.sleep(for: .milliseconds(33))
+            
+            alternateWindow.isHidden = true
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Application/Windowing/WindowManager.swift:73]: It should be fine, if the task is cancelled, the Task.sleep will throw so we won't hit this point. Although this comment has made me notice I didn't use the duration I added above 🙈

File: ElementX/Sources/Application/Windowing/WindowManager.swift [hunk 2]
```diff
@@ -58,21 +63,33 @@ class WindowManager {
...
+            
+            // We don't know what route the app will use when returning back
+            // to the main window, so end any editing operation now to avoid
+            // e.g. the keyboard being displayed on top of a call sheet.
+            mainWindow.endEditing(true)
+            
+            // Delay hiding to make sure the alternate window is visible.
+            try await Task.sleep(for: .milliseconds(33))
+            
+            overlayWindow.isHidden = true
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Application/Windowing/WindowManager.swift:90]: Ditto

File: UnitTests/Sources/AppLock/AppLockTimerTests.swift [hunk 3]
```diff
@@ -128,22 +128,58 @@ class AppLockTimerTests: XCTestCase {
...
+    
+    // MARK: - Helpers
+    
     /// Sets up the timer for testing.
     /// - Parameters:
     ///   - unlocked: Whether the timer should consider itself unlocked or not.
     ///   - backgroundedDate: If not nil, the timer will consider the app to have been backgrounded at the specified date.
-    private func setupTimer(unlocked: Bool, backgroundedAt backgroundedDate: Date? = nil) {
-        timer = AppLockTimer(gracePeriod: 180)
+    private func setupTimer(gracePeriod: TimeInterval = 180, unlocked: Bool, backgroundedAt backgroundedDate: Date? = nil) {
```
[the remark this answers is not in the document]
Comment (pixlwave) [reply UnitTests/Sources/AppLock/AppLockTimerTests.swift:181]: Ah yeah good point 👍

File: ElementX/Sources/Application/Windowing/WindowManager.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (alfogrillo) [reply ElementX/Sources/Application/Windowing/WindowManager.swift:73]: Looks good, my brain parsed `try?` on the sleep. 😅

[the remark this answers is not in the document]
Comment (alfogrillo) [reply ElementX/Sources/Application/Windowing/WindowManager.swift:73]: Looks good, my brain parsed `try?` on the sleep. 😅

**Pull request merged**